In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import StandardScaler
import joblib

In [2]:
rfm = pd.read_csv("../data/processed/rfm_dataset.csv")

rfm.head()

,CustomerID,Recency,Frequency,Monetary
0,12346.0,326,1,77183.60
1,12347.0,2,7,4310.00
2,12348.0,75,4,1797.24
3,12349.0,19,1,1757.55
4,12350.0,310,1,334.40


In [4]:
rfm_features = rfm[["Recency", "Frequency", "Monetary"]]

rfm_features.describe()

,Recency,Frequency,Monetary
count,4338.000000,4338.000000,4338.000000
mean,92.536422,4.272015,2048.688081
std,100.014169,7.697998,8985.230220
min,1.000000,1.000000,3.750000
25%,18.000000,1.000000,306.482500
50%,51.000000,2.000000,668.570000
75%,142.000000,5.000000,1660.597500
max,374.000000,209.000000,280206.020000


In [5]:
scaler = StandardScaler()

scaled_features = scaler.fit_transform(rfm_features)

In [6]:
scaled_rfm = pd.DataFrame(
    scaled_features,
    columns=rfm_features.columns
)

scaled_rfm.head()

,Recency,Frequency,Monetary
0,2.334574,-0.425097,8.363010
1,-0.905340,0.354417,0.251699
2,-0.175360,-0.035340,-0.027988
3,-0.735345,-0.425097,-0.032406
4,2.174578,-0.425097,-0.190812


In [7]:
scaled_rfm.describe().round(2)

,Recency,Frequency,Monetary
count,4338.00,4338.00,4338.00
mean,0.00,0.00,0.00
std,1.00,1.00,1.00
min,-0.92,-0.43,-0.23
25%,-0.75,-0.43,-0.19
50%,-0.42,-0.30,-0.15
75%,0.49,0.09,-0.04
max,2.81,26.60,30.96


In [8]:
joblib.dump(
    scaler,
    "../models/scaler.pkl"
)

print("Scaler saved successfully!")

Scaler saved successfully!


In [9]:
scaled_rfm.to_csv(
    "../data/processed/scaled_rfm.csv",
    index=False
)

print("Scaled dataset saved successfully!")

Scaled dataset saved successfully!


| Feature   | Before Scaling | After Scaling     |
| --------- | -------------- | ----------------- |
| Recency   | 1–374          | Mean ≈ 0, Std ≈ 1 |
| Frequency | 1–209          | Mean ≈ 0, Std ≈ 1 |
| Monetary  | 3.75–280206    | Mean ≈ 0, Std ≈ 1 |
